### PR TITLE
[Console] Add Application::runCommand() to run sub-commands easily

### PR DESCRIPTION
**Pull request by mnapoli**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

This PR adds a new `Application::runCommand()` method that allows to easily run a sub-command from within a command.

[How to run a sub-command today](http://symfony.com/doc/current/components/console/introduction.html#calling-an-existing-command):

``` php
protected function execute(InputInterface $input, OutputInterface $output)
{
    $command = $this->getApplication()->find('demo:greet');

    $arguments = array(
        'command' => 'demo:greet',
        'name'    => 'Fabien',
        '--yell'  => true,
    );

    $greetInput = new ArrayInput($arguments);
    $returnCode = $command->run($greetInput, $output);

    // ...
}
```

With this change:

``` php
protected function execute(InputInterface $input, OutputInterface $output)
{
    $app = $this->getApplication();
    $returnCode = $app->runCommand('demo:greet Fabien --yell', $output);

    // ...
}
```

I think that could be useful to avoid having to resort to use yet another tool to run grouped commands, for example Phing. Instead of running `bin/console db:drop --force && bin/console db:create && bin/console db:fixtures:load`, we could have a Symfony command that would run all that.

And yes usually you are better off calling services directly instead of sub-commands, but in some applications it's usually not worth it, especially micro-applications. For example I [ran into this in Silly](https://github.com/mnapoli/silly/issues/17):

``` php
$app->command('init', function ($input, $output) use ($app) {
    $app->runCommand('db:drop --force', $output)
    $app->runCommand('db:create', $output)
    $app->runCommand('db:fixtures', $output)
});
```

That looks super simple and easy to maintain.

Feedback welcome on the method and parameter names, I'm not 100% satisfied but it's better than nothing…

### CODE REVIEW

**Comment by fabpot** [comment]
:+1:

**Comment by Tobion** [comment]
Those before and after comparison does not seem valid as the new approach does not support correct escaping of command arguments as far as I see.

**Comment by mnapoli** [comment]
@Tobion just to be sure do you see that as a problem with the PR itself or just the example? Is there something that needs fixing?

File: src/Symfony/Component/Console/Application.php [hunk 1]
```diff
@@ -1018,6 +1020,24 @@ public function setDefaultCommand($commandName)
...
 
+    /**
+     * Helper to run a sub-command from a command.
+     *
+     * @param string               $command Command that should be run.
+     * @param OutputInterface|null $output  The output to use. If not provided, the output will be silenced.
+     *
+     * @return int 0 if everything went fine, or an error code
+     */
+    public function runCommand($command, OutputInterface $output = null)
```
Comment (Tobion) [inline src/Symfony/Component/Console/Application.php]: phpdoc is missing possible exceptions that can be thrown

**Comment by fabpot** [comment]
Indeed, @Tobion is right. To be fair, your "before" example should be:

```php
protected function execute(InputInterface $input, OutputInterface $output)
{
    $this->find('demo:greet')->run(new StringInput('demo:greet Fabien --yell'), $output);
}
```

which is not that much more complex than your after, so adding a method does not add much value.

**Comment by mnapoli** [comment]
Right, the amount of code is not so different, it's more about DX. Up to you, let me know.

**Comment by Tobion** [comment]
Also your implementation does not respect the "default command" logic when there is no command name given (which is part of the doRun implementation).

**Comment by Tobion** [comment]
What I find strange from a DX perspective is rather that one has to repeat the command name when executing a command directly. So to me, those lines should translate to the same thing:

```
$this->find('demo:greet')->run(new StringInput('demo:greet Fabien --yell'), $output);
$this->find('demo:greet')->run(new StringInput('Fabien --yell'), $output);
```

**Comment by mnapoli** [comment]
@Tobion it doesn't respect `--help` or `--version` either for example. This is a helper to run a sub-command, I think it makes sense but I might be wrong.

**Comment by Tobion** [comment]
It's a public method on the application. So you cannot just assume it's only used for "running a sub-command from a command". It's not bound to any command at all.

So to me it doesn't belong in the application class at all. It could be a separate class as all the information is already available through public methods.

**Comment by fabpot** [comment]
Closing as the new code is not that much shorter.

**Comment by mnapoli** [comment]
Fair enough, I'll add it to Silly directly (for those interested: https://github.com/mnapoli/silly/issues/17).

**Comment by mnapoli** [comment]
Stumbled upon that again today, I just want to correct the examples given in the thread: you can't call `$this->find(...`, you have to do `$this->getApplication()->find(…`.

Here is a complete example based from a real use case:

```
$this->getApplication()->find('assets:install')->run(new StringInput('assets:install assets --symlink'), $output);
```

The line is not really easy to read, and quite long (sometimes even longer), I still believe it should be improved, here is what it would look like with that PR:

```
$app->runCommand('assets:install assets --symlink', $output);
```

**Pull request closed without merging**